teps/draft-20200716-task-results-in-final-tasks.md

```diff
@@ -0,0 +1,304 @@
+all non-final tasks finishes.
+
+**Q. What happens to `Pipeline` when the dependent task `boskos-acquire` either failed or not executed and
+the task result `leased-resource` is not initialized?**
+
+**A.** The finally `Task` `boskos-release` is attempted but declared failure since the task result
+`leased-resource` is not initialized. Pipeline controller logs the reason of this failure as invalid result
+reference `InvalidTaskResultReference`. `Pipeline` continues executing rest of the final tasks and exits with failure.
+
+**Q. What happens when the dependent `Task` succeeds but the `Task` `Result` is empty?**
+
+**A.** It could be possible that the task author intentionally chooses to leave the `Task` `Result` empty
+(creating an empty file at `/tekton/results/`). `Pipeline` leaves that param value empty and executes
+the final task. Empty `Task` `Result` is considered a valid value and results in successful variable
+substitution. The error handling of such empty `Task` `Result` is left to the final `Task` author.
+
+**Q. Can a final `Task` produce `Task` `Results`?**
+
+**A.** Nope, this proposal does not enable final tasks to produce `Task` `Results`.
+
+
+**Q. Explain the overall workflow.**
+
+**A.**
+
+```
+                   ------------------------------------------
+                  |             Param Validation              |
+                  | (required params provided and their types)|
+                   ------------------------------------------
+                                    |
+                                    V
+                   -------------------------------------------
+                  | Retrieve Execution Queue (All Final Tasks)|
+                   -------------------------------------------
+                                    |
+                                    V
+                -------------------------------------------------
+               | Validate and Apply Results from Dependent Tasks |
+                -------------------------------------------------
+                                    | (if the validation fails, declare that task failure but
+                                    | continue executing rest of the final tasks)
+                                    V
+                    ------------------------------------
+                   | Execute the Task and Create TaskRun |
+                    ------------------------------------
+```
+
+## Advantages
+
+* No API changes.
+
+* Finally contract does not break and all final tasks are attempted.
+
+* Consistent across non-final and final tasks.
+
+* Finally tasks do not need to explicitly check if the Task they depend on ran or not
+
+* This proposal is extensible with a param validation proposal using Schema -
+[#1393](https://github.com/tektoncd/pipeline/issues/1393#issuecomment-558833526) by reading default from
+`Task` specification.
```

Comment:
would it make sense to remove this reference to schema as well? maybe this is a discussion for somewhere else, but this is how i would see this feature interacting with schema:

- Param defaults should always pass their own schema; if a task declares a default that does not pass it's schema, the Task creation should fail
- If we one day add Result defaults, when a Task fails, the default of the result could be used as the Result of that Task. If a Task wants to use that result as the value for one if its own params, then the Task validation will evaluate that value; if the default is invalid, the validation will fail, if it is valid, the Task can continue

Reply:
sure @bobcatfish, shall I drop the whole bullet point along with the example?

Reply:
never mind, i just dropped the reference and have the example as is.